### ISSUE TITLE
error 0x00000578

### ISSUE THREAD

**Issue by lhksoft** [wxWidgets/wxWidgets#22178]
**Describe the bug**
After the program has run, and terminated correctly I see following error in the logwindow inside MS-Studio
`C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\include\wx/msw/private.h(371): 'GetWindowRect' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\include\wx/msw/private.h(371): 'GetWindowRect' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\include\wx/msw/private.h(371): 'GetWindowRect' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\src\msw\window.cpp(2063): 'MoveWindow' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\include\wx/msw/private.h(371): 'GetWindowRect' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\include\wx/msw/private.h(371): 'GetWindowRect' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\include\wx/msw/private.h(371): 'GetWindowRect' failed with error 0x00000578 (Ongeldige vensteringang.).
C:\dev\vcpkg\buildtrees\wxwidgets\src\54f4f955a1-0d7c66a59e.clean\src\msw\window.cpp(2063): 'MoveWindow' failed with error 0x00000578 (Ongeldige vensteringang.).`

**Expected vs observed behaviour**
I gues this shouldn't happen. If I did something wrong an assertion should occur, or I should find some documentation what can and what cannot be done. Apparently it's not a big issue in the low-level code of wxWidgets while the program runs as expected. Then one should code some checks in the code so these errors won't occur.

**Patch or snippet allowing to reproduce the problem**
I allready found what was wrong, you can find it at [https://forums.wxwidgets.org/viewtopic.php?f=23&t=49038](https://forums.wxwidgets.org/viewtopic.php?f=23&t=49038)
The error occus in <wx/msw/private.h> at line 371. It was because I called wxComboBox:Clear() before creation of that control was done. By calling the clear-function after window has been created solved the issue. But because even with the error the program ran fine, it would be better - I think - to add some extra safety checks in the wxMSW code

My patch would be something like this :
`
inline RECT wxGetWindowRect(HWND hwnd)
{
    RECT rect = { 0, 0, 0, 0 };

    if ( hwnd && ::IsWindow(hwnd) )
        ::GetWindowRect(hwnd, &rect);

    return rect; // if invalid window, or window not yet created, just return an invalid rect
}
`

**Platform and version information**
 - wxWidgets version you use: 3.1.5
 - wxWidgets port you use: wxMSW and wxGTK -- This issue only happened in the wxMSW port
 - OS and its version: Microsoft Windows [Version 10.0.19044.1566], and Linux Mint 20.2
 - For wxGTK only:
   + GTK version: 3.24.20
   + Which GDK backend is used: don't know
   + If relevant, i.e. for appearance-related problems: -none-

**Comment by vadz** [comment]
I agree that there is a problem here and that calling `Clear()` shouldn't result in these debug errors, but I think the right thing to do would be to check whether the window was created in `Clear()` itself -- and fail with an assert in wxMSW (and also wxOSX) just as we already do in wxGTK. Alternatively, perhaps we could simply do nothing if `Clear()` is called before the window is created, as it somewhat makes sense -- but this should also be done in all ports, i.e. in `wxItemContainer::Clear()` itself.

Allowing calling this low-level functions with invalid window doesn't seem like a good idea to me, as it could hide other, completely unrelated, errors elsewhere.

**Comment by vadz** [comment]
BTW, to reproduce the problem it's enough to just insert this
```cpp

    auto *c = new wxComboBox;
    c->Clear();
    delete c;
```
snippet in any sample (e.g. minimal).

**Comment by lhksoft** [comment]
> I agree that there is a problem here and that calling `Clear()` shouldn't result in these debug errors, but I think the right thing to do would be to check whether the window was created in `Clear()` itself -- and fail with an assert in wxMSW.....

I aggree, but also disagree a bit. 
The thing I do might be too advanced for just basic developpers (I gues, most of wxWidgets users are basic) because I need to have some controls which are a bit more advanced than those available in wxWidgets. Nevermind, I share the code once all is finished. It's all opensource.

What I mean, I call Clear to get a completely fresh start. Later I have some function to do a 'StartOver', this would be after the control already is created. Then it wouldn't be a big deal.
I suggest one could add extra safty in low-level wxwidgets code. Whenever things are called something that needs stuff like the window-rect, in msw you can do the check that the hwnd is valid or not. And if it isn't, just skip that code. But ensure other things are cleared (re-initialised) anyhow.
Like, for example, why call MoveWindow through a call of Clear ?

After I posted my issue at the usersforum of wxWidgets, I wend deeper into this problem and in my own code, I added a 'bool m_bCreated'. At C'tor it will be false, as soon ::Create is successfully called it will be true. Things I do, like calling ::Clear, I'll check for the value of m_bCreated. This is some simple solution.

I've created my own classes of a Combobox and/or Listbox wich will have bitmaps (from wxImage, so one could use all kind of images) with text (in a readonly combo, the bitmap and text also is drawn in the textbox) and a Listbox with optionally checkboxes (use them, or not and in that case don't draw the checkboxes) with bitmaps and text.
I have a function to pre-initialize the contents that must come in these boxes, through a func called InitializeFromList. This would be called before creation of the control, because the size must be adjusted/initialised. To be able to have a dynamic list, later on one can call 'StartOver', which will call InitializeFromList again to get other items in the list.. StartOver is called from within InitializeFromList, in StartOver ::Clear is called... 
So in this case if m_bCreated is false, in StartOver ::Clear won't be called. In that way, I won't have any error 0x00000578 anymore.

**Comment by vadz** [comment]
You really don't need any `m_bCreated`, just checking if `m_hWnd`, returned by `GetHandle()`, is non-null is enough.

In fact, I'm almost sure that just
```diff
diff --git a/src/common/ctrlsub.cpp b/src/common/ctrlsub.cpp
index 2450d3b655..d9e4dfceed 100644
--- a/src/common/ctrlsub.cpp
+++ b/src/common/ctrlsub.cpp
@@ -91,6 +91,9 @@ wxItemContainer::~wxItemContainer()

 void wxItemContainer::Clear()
 {
+    if ( !GetHandle() )
+        return;
+
     if ( HasClientObjectData() )
     {
         const unsigned count = GetCount();
```
would be enough to fix the problem and make `Clear()` work the same (do nothing) in all ports if it's called before creating the window.

**Comment by lhksoft** [comment]
> You really don't need any `m_bCreated`, just checking if `m_hWnd`, returned by `GetHandle()`, is non-null is enough.

Working with the `HWND` isn't low-level? I didn't found that in wxGTK.
What I'm coding should be able to compile/link at both wxMSW and wxGTK.

When I use `GetHandle`, it will work in wxGTK too?

I'm not low-level coding the wxWidgets API, but if one of the coders could apply that fix in a future release, it would be great.

**Comment by vadz** [comment]
> When I use `GetHandle`, it will work in wxGTK too?

Yes.
 
> I'm not low-level coding the wxWidgets API, but if one of the coders could apply that fix in a future release, it would be great.

We need to decide whether we want this to silently do nothing or fail and test this.

**Comment by lhksoft** [comment]
Well, I modified my code.
I use `GetHandle()` to test whether `Create` already has been called, if GetHandle returns NULL I skip any calls to Clear (I use it in wxComboBox, wxOwnerDrawnComboBox and wxVListBox).

The routine that calls my own `StartOver()` func, which in turn calls `Clear` and other things to reset the whole control, gets called sometimes before and sometimes after the Control is created. 
Thanks for the workaround. No problems anymore..

But I think it would be best to insert in low-level code too, so if anyone in the future would call Clear before creation of the control, they don't come with the same error. But I leave that to you (authors of wxWidgets).
